### PR TITLE
General env

### PR DESCRIPTION
**Pull request by ykerus**
Made several functionalities more general, e.g. adjusting annealing time, I replaced hardcoded params with argparse, and allow the user to choose the environment - tested with small number of similar environments, so might still have bugs for very different environments, though I changed the Q-net architecture to fit with the given environment.

### CODE REVIEW

File: scripts/DQN_main.py [hunk 1]
```diff
@@ -116,6 +137,9 @@ def main():
 
     parser.add_argument('--n_episodes', '-ne', type=int, default=100, help="Number of episodes to train model.")
     parser.add_argument('--batch_size', '-bs', type=int, default=64, help="Number of historical states to batch train with for each present state.")
+    parser.add_argument('--min_eps', type=int, default=0.05, help="Minimum epsilon after annealing.")
+    parser.add_argument('--max_eps', type=int, default=1, help="Maximum epsilon before annealing.")
+    parser.add_argument('--anneal_time', type=int, default=1000, help="Number of steps before reaching eps_min.")
```
Comment (DanielNobbe) [inline scripts/DQN_main.py]: Would be nice to add a short command here, like I did with the others

File: scripts/DQN_policy.py [hunk 2]
```diff
@@ -36,16 +34,16 @@ def sample_action(self, obs):
...
-            return action
+        # This piece of code was copied and altered from our version of lab4 - Yke
+        with torch.no_grad():
+            obs_tensor = torch.tensor(obs).type(torch.FloatTensor)
+            model_output = self.Q(obs_tensor)
+            a_greedy = torch.argmax(model_output).item()
+        probs = [self.epsilon / self.num_actions for i in range(self.num_actions)]
+
+        probs[a_greedy] += 1 - self.epsilon
+        assert sum(probs) == 1, "Probabilies should sum to 1"
```
Comment (DanielNobbe) [inline scripts/DQN_policy.py]: Change this to np.allclose(probs, 1). Sometimes the sum is 0.99999 and that should be okay.

File: scripts/DQN_main.py [hunk 3]
```diff
@@ -69,9 +76,9 @@ def run_episodes(train, Q, policy, memory, env, num_episodes, batch_size, discou
 
             if done:
                 if i % 10 == 0:
-                    print("{2} Episode {0} finished after {1} steps"
-                          .format(i, steps, '\033[92m' if steps >= 195 else '\033[99m'))
-                    print("epsilon: ", policy.epsilon)
+                    # loss and rewards are avg loss and reward per step
+                    print("[{:<4} done: step {:<5}| loss: {:<8.5} | rewards: {:8<.5} | eps: {:<6.5}"
```
Comment (DanielNobbe) [inline scripts/DQN_main.py]: Not sure how useful this reward value is, since it's 1.0 for most environments.

**Review by DanielNobbe — CHANGES_REQUESTED** [review]

**Review by DanielNobbe — APPROVED** [review]
